src/text_to_binary/grouped.ml

```diff
@@ -103,13 +103,31 @@ let add_mem value (fields : t) (curr : curr) =
   incr curr.mem;
   { fields with mem = Indexed.return index value :: fields.mem }
 
+let rec extract_block_types expr =
+  let aux instr =
+    match instr with
+    | Block (_str_opt, bt, expr1) ->
+      Option.to_list bt @ extract_block_types expr1
+    | Loop (_str_opt, bt, expr1) ->
+      Option.to_list bt @ extract_block_types expr1
+    | If_else (_str_opt, bt, expr1, expr2) ->
+      Option.to_list bt @ extract_block_types expr1 @ extract_block_types expr2
+    | Return_call_indirect (_ind, bt) -> [ bt ]
+    | Return_call_ref bt -> [ bt ]
+    | Call_indirect (_ind, bt) -> [ bt ]
```

Comment:
Can you merge these three cases into one ? (You'll have to rename `_ind` to `_`)